core/src/labels/labels.cpp

```diff
@@ -84,98 +69,221 @@ void Labels::update(const View& _view, float _dt, const std::vector<std::unique_
-                    if (!mesh0) { continue; }
-                    const auto& m1 = t1->getMesh(*style);
-                    if (!m1) { continue; }
-                    const LabelMesh* mesh1 = static_cast<const LabelMesh*>(m1.get());
-
-                    for (auto& l0 : mesh0->getLabels()) {
-                        if (!l0->canOcclude()) { continue; }
-
-                        for (auto& l1 : mesh1->getLabels()) {
-                            if (!l1 || !l1->canOcclude() || l0->hash() != l1->hash()) {
-                                continue;
-                            }
-                            float d2 = glm::distance2(l0->transform().state.screenPos,
-                                    l1->transform().state.screenPos);
-
-                            // The new label lies within the circle defined by the bbox of l0
-                            if (sqrt(d2) < std::max(l0->dimension().x, l0->dimension().y)) {
-                                l0->skipTransitions();
-                            }
+                        // The new label lies within the circle defined by the bbox of l0
+                        if (sqrt(d2) < std::max(l0->dimension().x, l0->dimension().y)) {
+                            l0->skipTransitions();
                         }
                     }
                 }
             }
         }
     }
+}
 
-    //// Update label meshes
+void Labels::checkRepeatGroups(std::vector<TextLabel*>& _visibleSet) const {
+    struct GroupElement {
+        glm::vec2 position;
+        float threshold;
+        const std::string* group;
+
+        bool operator==(const GroupElement& _ge) {
+            return _ge.position == position
+                && _ge.threshold == threshold
+                && *_ge.group == *group;
+        };
+    };
+
+    std::map<size_t, std::vector<GroupElement>> repeatGroups;
+
+    for (TextLabel* textLabel : _visibleSet) {
+        auto& options = textLabel->options();
+        if (options.repeatDistance == 0.f) { continue; }
+
+        size_t hash = std::hash<std::string>()(options.repeatGroup);
+
+        GroupElement element {
+            textLabel->center(),
+            options.repeatDistance,
+            &options.repeatGroup
+        };
+
+        auto it = repeatGroups.find(hash);
+        if (it == repeatGroups.end()) {
+            repeatGroups[hash].push_back(element);
+            continue;
+        }
+
+        std::vector<GroupElement>& group = it->second;
+        if (std::find(group.begin(), group.end(), element) != group.end()) {
+            // TBD: when is this case possible?
+            // -> When two tiles contain the same label?
+            continue;
+        }
+
+        bool add = true;
+        float threshold2 = pow(element.threshold, 2);
+
+        for (const GroupElement& ge : group) {
+            float d2 = distance2(ge.position, element.position);
+            if (d2 < threshold2) {
```

Comment:
Here one would need to check here if group.group == element.group due to possible hash collisions